connectors/sources/servicenow.py

```diff
@@ -0,0 +1,332 @@
+        """Setup the connection to the ServiceNow instance.
+
+        Args:
+            configuration (DataSourceConfiguration): Instance of DataSourceConfiguration class.
+        """
+
+        super().__init__(configuration=configuration)
+        self.concurrent_downloads = self.configuration["concurrent_downloads"]
+        self.servicenow_client = ServiceNowClient(configuration=configuration)
+
+    def tweak_bulk_options(self, options):
+        """Tweak bulk options as per concurrent downloads support by ServiceNow
+
+        Args:
+            options (dict): Config bulker options.
+        """
+
+        options["concurrent_downloads"] = self.concurrent_downloads
+
+    @classmethod
+    def get_default_configuration(cls):
+        return {
+            "url": {
+                "label": "Service URL",
+                "order": 1,
+                "type": "str",
+                "value": "http://127.0.0.1:9318",
+            },
+            "username": {
+                "label": "Username",
+                "order": 2,
+                "type": "str",
+                "value": "admin",
+            },
+            "password": {
+                "label": "Password",
+                "sensitive": True,
+                "order": 3,
+                "type": "str",
+                "value": "changeme",
+            },
+            "services": {
+                "display": "textarea",
+                "label": "Comma-separated list of services",
+                "order": 4,
+                "type": "list",
+                "value": "*",
+            },
+            "retry_count": {
+                "default_value": RETRIES,
+                "display": "numeric",
+                "label": "Retries per request",
+                "order": 5,
+                "required": False,
+                "type": "int",
+                "ui_restrictions": ["advanced"],
+                "value": RETRIES,
+            },
+            "concurrent_downloads": {
+                "default_value": MAX_CONCURRENCY_SUPPORT,
+                "display": "numeric",
+                "label": "Maximum concurrent downloads",
+                "order": 6,
+                "required": False,
+                "type": "int",
+                "ui_restrictions": ["advanced"],
+                "value": MAX_CONCURRENCY_SUPPORT,
+            },
+        }
+
+    async def _remote_validation(self):
+        """Validate configured services
+
+        Raises:
+            ConfigurableFieldValueError: Unavaliable services error.
+        """
+
+        if self.servicenow_client.services != ["*"]:
+            (
+                service_names,
+                invalid_services,
+            ) = await self.servicenow_client.filter_services()
+            if invalid_services:
+                raise ConfigurableFieldValueError(
+                    f"Services '{', '.join(invalid_services)}' are not available. Available services are: '{', '.join(service_names)}'"
+                )
```

Comment:
> It will be helpful in get_docs for preparing URLs with name & kept label for invalid services so that user can easily find out which entered services are invalid

In a comment above we were discussing names vs labels, this error here is an example of what is confusing me. It tells the user two things:
- Service _labels_ x, y, and z are unavailable
- Service _names_ a, b, and c are available 

It's comparing different things (labels vs names), so I think it can lead to confusion. Can you rework this error message to be consistent?
For example, either:

`f"Services '{', '.join(invalid_service_labels)}' are not available. Available services are: '{', '.join(service_labels)}'"`

Or

`f"Services '{', '.join(invalid_service_names)}' are not available. Available services are: '{', '.join(service_names)}'"`

Reply:
Got it 👍 Updated